plugins/signals/src/api/SignalClient.ts

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { SignalApi } from '@backstage/plugin-signals-react';
+import { JsonObject } from '@backstage/types';
+import { DiscoveryApi, IdentityApi } from '@backstage/core-plugin-api';
+import { v4 as uuid } from 'uuid';
+
+type Subscription = {
+  topic: string;
+  callback: (message: JsonObject) => void;
+};
+
+const WS_CLOSE_NORMAL = 1000;
+const WS_CLOSE_GOING_AWAY = 1001;
+
+/** @public */
+export class SignalClient implements SignalApi {
+  static readonly DEFAULT_CONNECT_TIMEOUT_MS: number = 1000;
+  static readonly DEFAULT_RECONNECT_TIMEOUT_MS: number = 5000;
+  private ws: WebSocket | null = null;
+  private subscriptions: Map<string, Subscription> = new Map();
+  private messageQueue: string[] = [];
+  private reconnectTo: any;
+
+  static create(options: {
+    identity: IdentityApi;
+    discoveryApi: DiscoveryApi;
+    connectTimeout?: number;
+    reconnectTimeout?: number;
+  }) {
+    const {
+      identity,
+      discoveryApi,
+      connectTimeout = SignalClient.DEFAULT_CONNECT_TIMEOUT_MS,
+      reconnectTimeout = SignalClient.DEFAULT_RECONNECT_TIMEOUT_MS,
+    } = options;
+    return new SignalClient(
+      identity,
+      discoveryApi,
+      connectTimeout,
+      reconnectTimeout,
+    );
+  }
+
+  private constructor(
+    private identity: IdentityApi,
+    private discoveryApi: DiscoveryApi,
+    private connectTimeout: number,
+    private reconnectTimeout: number,
+  ) {}
+
+  subscribe(topic: string, onMessage: (message: JsonObject) => void): string {
+    const subscriptionId = uuid();
+    const exists = [...this.subscriptions.values()].find(
+      sub => sub.topic === topic,
+    );
+    this.subscriptions.set(subscriptionId, { topic, callback: onMessage });
+
+    this.connect()
+      .then(() => {
```

Comment:
Handling subscriptions, topics and connections can get fairly complex pretty quick. I think it might already now be worth breaking you the WebSocket management into some form of transport abstraction. It can also take care of queueing unsent messages and whatnot as well, maybe even deduplicating incoming messages and resending. May also be worth setting up a request/response scheme for the messages coming from the frontend as well since subscriptions are a bit more important to get an ack for.

Not in this PR though, just a potential future refactor.